### ISSUE TITLE
Replace or remove Cheetah

### ISSUE THREAD

**Issue by ericvaandering** [dmwm/WMCore#6460]
Cheetah has not been updated since 2010 and is not python 3 compatible. @ticoann I see its used in ReqMgr and is referenced in WMCore/WebTools. Will it be used after ReqMgr is removed in favor of RM2?

**Comment by ticoann** [comment]
We don't use Cheetah template in WMCore other than ReqMgr, however I am not I am not sure whether other applications using WebTools use Cheetah, DBS, maybe DAS. I will check with Yuyi and Valentin, (I am not sure whether Crab uses WebTools.

**Comment by mapellidario** [comment]
A possible solution could be using Cheetah3, https://pypi.org/project/Cheetah3/, which supports both py2.7 and python3.

I followed the instructions at https://github.com/dmwm/WMCore/wiki/Docker-and-WMCore#install-python-packages-via-pip to install Cheetah3 via pip in my docker container.
They did not change the name of the module (Cheetah3 is imported with `import Cheetah` in both py2 and py3), so that I had to move the directory `./unittestdeploy/wmagent/1.3.2.pre2/sw/slc7_amd64_gcc630/external/py2-cheetah/2.4.0-comp2/lib/python2.7/site-packages/Cheetah` so that python finds the new version of cheetah, the one installer by Cheetah3.

I checked that python finds the right version with

```python
[root@82f179d8a51e WMCore]# python 
>>> import Cheetah
>>> Cheetah.Version
'3.2.5'
```

I then looked for the modules that use Cheetah:

```bash
> grep -ri "cheetah"
src/templates/WMCore/WebTools/ConfigFile.tmpl:                                  <p><b>templates</b> is the default directory for the applications Cheetah
src/templates/WMCore/WebTools/Page.tmpl:                                        href="documentation/ConfigFile">ConfigFile</a>) and use Cheetah to 
src/python/WMCore/ReqMgr/Web/tools.py:    from Cheetah.Template import Template
src/python/WMCore/ReqMgr/Web/tools.py:    TemplatedPage is a class that provides simple Cheetah templating
src/python/WMCore/ReqMgr/Web/tools.py:            templates = 'Cheetah'
src/python/WMCore/ReqMgr/Web/tools.py:        return self.templatepage_cheetah(ifile, *args, **kwargs)
src/python/WMCore/ReqMgr/Web/tools.py:    def templatepage_cheetah(self, ifile=None, *args, **kwargs):
src/python/WMCore/WebTools/Welcome.py:from Cheetah import Version as cheetah_version
src/python/WMCore/WebTools/Page.py:from Cheetah import Version
src/python/WMCore/WebTools/Page.py:from Cheetah.Template import Template
src/python/WMCore/WebTools/Page.py:    TemplatedPage is a class that provides simple Cheetah templating
src/python/WMCore/WebTools/Page.py:        self.debug("Using Cheetah version: %s" % Version)
src/python/WMCore/WebTools/Page.py:        Apply the cheetah template specified by file to the data in arg and kwargs.
requirements.txt:Cheetah==2.4.0
requirements.wmagent.txt:Cheetah==2.4.0
```

I did not find any unit test directly related to the modules that usa cheetah, but `src/python/WMCore/WebTools/RESTFormatter.py` imports `from WMCore.WebTools.Page import TemplatedPage, _setCherryPyHeaders` so that it seems to make use of cheetah.
I run its unit test and everything seems fine

```bash
python setup.py test --buildBotMode=true --reallyDeleteMyDatabaseAfterEveryTest=true --testCertainPath=test/python/WMCore_t/WebTools_t/RESTFormat_t.py

running test
Using the tests below: test/python/WMCore_t/WebTools_t/RESTFormat_t.py
#### WE ARE DELETING YOUR DATABASE. 3 SECONDS TO CANCEL ####
#### buildbotmode is true
We are going to trash databases after every test
### We are in buildbot mode ###
----------------------------------------------------------------------
Ran 5 tests in 0.001s

OK
path lists is ['test/python/WMCore_t/WebTools_t/RESTFormat_t.py']
Out of 5 cases, we will run 5
#1 testGenerator (WMCore_t.WebTools_t.RESTFormat_t.RESTFormatTest) ... CherryPy Checker:
The use of 'localhost' as a socket host can cause problems on newer systems, since 'localhost' can map to either an IPv4 or an IPv6 address. You should use '127.0.0.1' or '[::1]' instead.

CherryPy Checker:
The Application mounted at '/unittests/rest' has an empty config.

ok
#2 list1 takes no arguments, it should raise an error if called with one. Require json output. ... ok
#3 testReturnFormat (WMCore_t.WebTools_t.RESTFormat_t.RESTFormatTest) ... ok
#4 testSupportedFormat (WMCore_t.WebTools_t.RESTFormat_t.RESTFormatTest) ... ok
#5 testUnsupportedFormat (WMCore_t.WebTools_t.RESTFormat_t.RESTFormatTest) ... ok

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 5 tests in 0.833s

OK
Testing complete, there are now 1 threads
/home/dmwm/unittestdeploy/wmagent/1.3.2.pre2/sw/slc7_amd64_gcc630/external/py2-cherrypy/17.4.0/lib/python2.7/site-packages/cherrypy/process/wspbus.py:258: RuntimeWarning: The main thread is exiting, but the Bus is in the states.STOPPED state; shutting it down automatically now. You must either call bus.block() after start(), or call bus.exit() before the main thread exits.
  'main thread exits.' % self.state, RuntimeWarning)
```

This is promising. I can not think of any better way to test this, do you have idea? Should we try to change a spec file and let jenkins do its job?

**Comment by amaltaro** [comment]
Dario, 4 years ago we had this PR to support jinja templates (which AFAIK supports both major python versions):
https://github.com/dmwm/WMCore/pull/7100

If Cheetah now supports python3 - AND it would not require any code changes from our side neither from the other services depending on WMCore - then integrating Cheetah3 into our stack is certainly the best way to go. We can even consider dropping Jinja templates IMO.

Of course we would have to properly test this cheetah3, but if it works fine. Then it would be just a matter of updating the spec files of anything depending on py2-cheetah to py2-cheetah3.

What do you think?

**Comment by mapellidario** [comment]
I do agree with you. I would only change our spec files and test the new Cheetah3. I would leave any refactoring, i.e. dropping jinja templates, after the migration is complete

**Comment by amaltaro** [comment]
@mapellidario as we discussed in the past week, I'm setting this issue free such that others can pick it, if they want.

**Comment by yuyiguo** [comment]
Hi All,

I am going through all DBS external libraries. Cheetah is used in by DBS by importing WMCore.WebTools.Page. It looks to me that cheetah3 is a replacement to cheetah.

**Comment by khurtado** [comment]
@yuyiguo : The [PR](https://github.com/cms-sw/cmsdist/pull/6898) from @mapellidario with the change suggestion I put at the end builds for me.

**Comment by khurtado** [comment]
@yuyiguo The suggestion was just pushed today in the [PR](https://github.com/cms-sw/cmsdist/pull/6898). Could you please test when you get the chance?

**Comment by amaltaro** [comment]
This is supposed to have been fixed by: https://github.com/cms-sw/cmsdist/pull/6898

Thanks to Dario and Kenyi. From my side, this issue is ready to be closed. Please feel free to do so.

**Comment by mapellidario** [comment]
Yes, it's fine closing this issue for me, too. In case we find any problem we can always open a new issue. (I leave to you the dopamine rush of pressing the "close with comment" button ;) )

**Comment by amaltaro** [comment]
If you insist :-D
Summary: our COMP python3 stack supports `Cheetah3`, which **should** have the same interface/import as the `Cheetah` package.

Closing this issue.
